### PR TITLE
SCSS Variables: Updated src variables file and added customizable variables for Tabs plugin

### PR DESCRIPTION
**Pull request by melodyma**
- Updated formatting of src variables file for easier navigation and customization
  - Ordering of overridden Bootstrap variables is the same as in Bootstrap's variable file
  - WET plugin variables are ordered alphabetically by plugin name
  - Added instructions for overriding Bootstrap variables not already in src variables
- Added customizable variables for Tabs plugin (i.e. colours, border radiuses)
  - Reduced number of greys for tab border

### CODE REVIEW

File: src/_variables.scss [hunk 1]
```diff
@@ -3,103 +3,222 @@
...
-$mm-progress-bg-color: #fff !default;
-$mm-cc-fg-color: #fff !default;
-$mm-cc-bg-color: #000 !default;
-$mm-ctrl-fg-color: #fff !default;
-$mm-ctrl-bg-color: #3e3e3e !default;
-$mm-ctrl-outline-color: #4aafff !default;
-$mm-ctrl-cc-color: #4aafff !default;
-$mm-volume-bg-color: #aaa !default;
-$mm-volume-thumb-color: #fff !default;
+//== Media queries breakpoints
```
Comment (LaurentGoderre) [inline src/_variables.scss]: These are custom breakpoints that are not defined in BootStrap , I think the text should reflect that

File: src/_variables.scss [hunk 2]
```diff
@@ -3,103 +3,222 @@
...
+//
+//##
+
+$spacing:                 .375em !default;
+$ddRtnWidth:              3.375em !default;
+
+
+//== Multimedia player
+//
+//##
```
Comment (LaurentGoderre) [inline src/_variables.scss]: What does this mean?

Comment (melodyma) [inline src/_variables.scss]: The `//##`? Those are just placeholders for comments for formatting consistency similar to the Bootstrap variables file. We can take them out if you'd like.

File: src/_variables.scss (hunk 1 above)
Comment (melodyma) [inline src/_variables.scss]: Yes, I can make this change. Is this the only change outstanding? If so, I can do a fix and commit. Otherwise, I'll wait for your comments to come if there are any more. Thanks.

File: src/_variables.scss [hunk 3]
```diff
@@ -3,103 +3,222 @@
...
+$carousel-tabpanel-figcaption-color:         #fff !default;
+$carousel-tabpanel-figcaption-link-color:    $carousel-tabpanel-figcaption-color !default;
+$carousel-tabpanel-figcaption-bg:            $clrDrabBlue !default;
+
+// Carousel Style 2 with Thumbnails
+$carousel-s2-tablist-controls-bg:            $clrLight !default;
+
+$carousel-s2-tablist-controls-color:         $clrDrabBlue !default;
+$carousel-s2-tablist-controls-button-bg:     #fff !default;
+$carousel-s2-tablist-controls-button-shadow: $clrDrabBlue !default;
```
Comment (LaurentGoderre) [inline src/_variables.scss]: I am wondering if this should be changed to the whole shadow so that you can turn it off completely if you want

```
$carousel-s2-tablist-controls-button-shadow: 0 0 4px $clrDrabBlue !default;
```

File: src/_variables.scss [hunk 4]
```diff
@@ -3,103 +3,222 @@
...
+$carousel-s2-tablist-controls-button-radius: 999px !default;
+
+$carousel-s2-tablist-thumbnail-active-border: #666 !default;
```
Comment (LaurentGoderre) [inline src/_variables.scss]: I am wondering if this can also be the full border value to add more control

File: src/_variables.scss [hunk 5]
```diff
@@ -3,103 +3,222 @@
...
+$mm-volume-thumb-color:   #fff !default;
+
+
+//== Tabs
+//
+//##
+
+// Tabbed Interface
+$tablist-color:                  #000 !default;
+$tablist-border:                 $clrMedium !default;
```
Comment (LaurentGoderre) [inline src/_variables.scss]: I am wondering if this can also be the full border value to add more control

File: src/_variables.scss [hunk 6]
```diff
@@ -3,103 +3,222 @@
...
+$tabpanel-border:                $tablist-border !default;
```
Comment (LaurentGoderre) [inline src/_variables.scss]: I am wondering if this can also be the full border value to add more control

File: src/_variables.scss [hunk 7]
```diff
@@ -3,103 +3,222 @@
...
+$tablist-bg:                     #ebf2fc !default;
+$tablist-hover-bg:               $clrMedium !default;
+$tablist-active-bg:              #fff !default;
+$tablist-active-link-border-top: #666 !default;
+// Tablist at sm max && xxs max
+$tablist-screen-sm-details-border:           $tablist-border !default;
```
Comment (LaurentGoderre) [inline src/_variables.scss]: I am wondering if this can also be the full border value to add more control

File: src/_variables.scss [hunk 8]
```diff
@@ -3,103 +3,222 @@
...
+//##
+
+// Tabbed Interface
+$tablist-color:                  #000 !default;
+$tablist-border:                 $clrMedium !default;
+$tabpanel-border:                $tablist-border !default;
+$tablist-bg:                     #ebf2fc !default;
+$tablist-hover-bg:               $clrMedium !default;
+$tablist-active-bg:              #fff !default;
+$tablist-active-link-border-top: #666 !default;
```
Comment (LaurentGoderre) [inline src/_variables.scss]: This seems very specific....is there a way to make this more generic and controllable?

Comment (LaurentGoderre) [inline src/_variables.scss]: This article mentions that by splitting it up under 3 different properties you can specify different values for each edge

border {
    width: 4px 0 0 0;
    style: solid;
    color: #666;
}

**Comment by melodyma** [comment]
@LaurentGoderre - The Travis CI build failed because of a bundle install issue for rake. It doesn't seem like it was anything to do with my commit. I did a `grunt lint` on my machine and everything passed. How can I run the Travis build again to make this go through? Thanks.

**Comment by LaurentGoderre** [comment]
I can restart builds when you need it but it seems the build for this PR did pass...

**Comment by melodyma** [comment]
Perfect. I've made my changes based on your comments and it's ready for review. Let me know if there are any other issues.

**Comment by melodyma** [comment]
@LaurentGoderre - When do you think you'd be able to review and merge this? I'd like to start working on doing the same for the other plugins. Thanks!

**Comment by LaurentGoderre** [comment]
Can you squash the last two commits?

**Comment by LaurentGoderre** [comment]
On thing I noticed, and it's not your code, is that there is no consistency in the way the colors are handled. We should pick one and stick with it. 

What would you propose? Here are some possible options

* `$purpose-fg-color, $purpose-bg-color`
* `$purpose-fg, $purpose-bg`
* `$purpose-color, $purpose-background`

File: src/_variables.scss [hunk 9]
```diff
@@ -3,103 +3,220 @@
...
+
+//== Calendar
+//
+//##
+
+$cal-text-head:           #fff !default;
+$cal-text-body:           #000 !default;
+
+$cal-dark:                #333 !default;
+$cal-medium:              #555 !default;
```
Comment (LaurentGoderre) [inline src/_variables.scss]: I know these are not from your but they should be changed to reflected their purpose, not their styles

**Comment by nschonni** [comment]
Awesome stuff :heart: :heart: :heart: , but I don't think the variable values should be aligned. This will cause noise (and muddy git-blame) every time a longer variable name is added the file.

**Comment by nschonni** [comment]
@LaurentGoderre SCSSLint will have some stuff to help in this exercise in their next release https://github.com/causes/scss-lint/issues/337

**Comment by LaurentGoderre** [comment]
@nschonni do you mean aligned in characters? Btw @melodyma where did you get this convention

**Comment by nschonni** [comment]
@LaurentGoderre yeah, I was talking about the alignment https://github.com/causes/scss-lint/blob/master/lib/scss_lint/linter/README.md#spaceafterpropertycolon
It is a fairly common format (https://github.com/twbs/bootstrap/blob/master/less/variables.less) that some IDEs apply by default

**Comment by melodyma** [comment]
@LaurentGoderre - I put it in because it's a file that's meant for external consumption, so it's easier to read and navigate for other users when they are using the file. Bootstrap also uses this convention for their variables file. However, @nschonni makes a good point. I'm happy to change it since it's just a personal preference.

**Comment by nschonni** [comment]
@melodyma maybe that style could be applied once all the "variablization" of the plug-ins is finished.

**Comment by melodyma** [comment]
@LaurentGoderre - For colours, it would be most consistent if all the variables were `$*-color`. Bootstrap eliminates it for `bg`, but because we have `fg`, `cc` and everything in between and beyond, it may be wise to override the Bootstrap defaults to something that is more semantically maintainable for us. However, the downside to this approach is that we'll have to add more code to override the defaults and it may not be obvious to others what's happening right away. Perhaps, we can also meet somewhere in the middle and apply `$*-color` only if it's not a `$*-bg` or `$*-fg`. Thoughts?

@nschonni - I concur with your approach!

**Comment by LaurentGoderre** [comment]
I want what is the clearest and it should maybe be clear for people that are not familiar to CSS as well so explicitly specifying which type of color might be desirable. What do you think?

**Comment by LaurentGoderre** [comment]
We could also use `text-color` `border-color` and `bg-color` to be more explicit

**Comment by melodyma** [comment]
Should all the Bootstrap overrides also be appended with `-color` then?

**Comment by LaurentGoderre** [comment]
It depends what we decide on...

**Comment by melodyma** [comment]
My take is that adding `-color` will be more usable.

**Comment by LaurentGoderre** [comment]
And how do you difference the different color type?

**Comment by melodyma** [comment]
Sorry, could you give an example of different colour types?

**Comment by LaurentGoderre** [comment]
text, border and background

**Comment by melodyma** [comment]
I think if we're going to add `-color`, we might as well do it for all colour-associated variables. The borders one will have to have the color attribute split out from the rest of the attributes.

**Comment by LaurentGoderre** [comment]
Can you apply this convention to all our plugin variables (not the Bootstrap ones)

**Comment by melodyma** [comment]
@LaurentGoderre - Cleaned up this PR as per your comments. Some notes:

* I couldn't get more specific with the calendar colours based on how the variables were extended in the plugin. May be worthwhile to look at how the calendar plugin variables/mixins/css were organized in a separate dedicated PR.
* The border attributes had to be separated into three separate attributes each (width, style, color). The snippet you provided earlier doesn't compile properly and I had to explicitly spell each attribute out as `border-[attribute]`
* Not sure why the Travis build failed again. Seem like an npm package error. I ran a grunt dist locally that showed that my PR is clean.

That's all!

**Comment by LaurentGoderre** [comment]
The build is failing because of the sass linting (which is not available in grunt/node.js). For our project we use tabs instead of space so you need to configure your editor to use tabs for indenting

File: src/_variables.scss [hunk 10]
```diff
@@ -73,33 +138,100 @@ $pagination-hover-color: $btn-default-color !default;
...
+
+$tablist-active-link-border-color: #666 !default;
+$tablist-active-link-border-style: solid !default;
+$tablist-active-link-border-width: 4px 0 0 0 !default;
+
+// Tablist at sm max && xxs max
+$tablist-screen-sm-details-border-color: $tablist-border-color !default;
+$tablist-screen-sm-details-border-style: solid !default;
+$tablist-screen-sm-details-border-width: 0 1px 1px 1px !default;
+$tablist-screen-sm-details-top-border-width: 1px !default;
```
Comment (LaurentGoderre) [inline src/_variables.scss]: This probably need to be expanded to allow different radius on all edges

File: src/plugins/tabs/_screen-sm-max.scss [hunk 11]
```diff
@@ -9,18 +9,29 @@
...
+				}
 			}
 
 			&:last-of-type {
-				border-bottom-left-radius: 4px;
-				border-bottom-right-radius: 4px;
+				border: {
+					bottom: {
+						left-radius: $tablist-screen-sm-details-border-radius;
+						right-radius: $tablist-screen-sm-details-border-radius;
```
Comment (LaurentGoderre) [inline src/plugins/tabs/_screen-sm-max.scss]: We should investigate if we can remove the bottom so that this can be applied to all edges

File: src/_variables.scss (hunk 10 above)
Comment (melodyma) [inline src/_variables.scss]: @LaurentGoderre - Realistically, the top left and right radii will be different from the bottom left and right radii, but left and right radii would never be different from each other. Does it make sense to split them up to 4 separate radii or two pairs? Trying to balance the amount of customization.

File: src/plugins/tabs/_screen-sm-max.scss (hunk 11 above)
Comment (melodyma) [inline src/plugins/tabs/_screen-sm-max.scss]: I think this ties into my comment above.

File: src/_variables.scss (hunk 10 above)
Comment (LaurentGoderre) [inline src/_variables.scss]: Well in the example above `$tablist-screen-sm-details-top-border-width` is not needed since `$tablist-screen-sm-details-border-width` can specify the width on each edge. Can't the same be done for the radius?

```
$tablist-screen-sm-details-border-radius: 5px 0 5px 0;
```

**Comment by melodyma** [comment]
@LaurentGoderre - OK. I think this is good now. I had to refactor the `tabs_screen-sm-max.scss` file to get the borders to show in the right places so that you can control the borders as variables. Let me know if there's anything else!

**Comment by LaurentGoderre** [comment]
@melodyma sounds good. We should clean it up ASAP before people actually start using those variables ;)

File: src/plugins/calendar/_base.scss [hunk 12]
```diff
@@ -132,7 +132,6 @@
 .cal-days {
 	a {
 		@extend %cal-days-a-div;
-		color: $cal-days-link;
```
Comment (LaurentGoderre) [inline src/plugins/calendar/_base.scss]: Why was this removed?

Comment (melodyma) [inline src/plugins/calendar/_base.scss]: It was declared, but wasn't actually used. The links were overrode by the styles from `.cal-wd`. I think the entire calendar plugin needs to be reviewed independently in another PR like the tabs plugin. For this PR, I was thinking of just keeping the variables that were there already that were actively used by the plugin.

screen shot 2015-02-13 at 5 55 47 am

Comment (LaurentGoderre) [inline src/plugins/calendar/_base.scss]: Ok, thanks for the explanation!
